OpenRA.Game/Network/Nat.cs

```diff
@@ -0,0 +1,91 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Mono.Nat;
+
+namespace OpenRA.Network
+{
+	public enum NatStatus { Enabled, Disabled, NotSupported }
+
+	public class Nat
+	{
+		public NatStatus Status
+		{
+			get
+			{
+				return NatUtility.IsSearching ? natDevice != null ?
+					NatStatus.Enabled : NatStatus.NotSupported : NatStatus.Disabled;
+			}
+		}
+
+		Mapping mapping;
+		INatDevice natDevice;
+
+		public Nat()
+		{
+			if (Game.Settings.Server.DiscoverNatDevices)
+			{
+				NatUtility.DeviceFound += DeviceFound;
+				NatUtility.StartDiscovery();
+			}
+		}
+
+		readonly SemaphoreSlim locker = new SemaphoreSlim(1, 1);
+
+		async void DeviceFound(object sender, DeviceEventArgs args)
+		{
+			await locker.WaitAsync();
+			try
+			{
+				// Only interact with one at a time. Some support both UPnP and NAT-PMP.
+				natDevice = args.Device;
+
+				Log.Write("nat", "Device found: {0}", natDevice.NatProtocol);
+				Log.Write("nat", "Type: {0}", natDevice.GetType().Name);
+			}
+			finally
+			{
+				locker.Release();
+			}
+		}
+
+		public async Task ForwardPort(int listen, int external)
+		{
+			var lifetime = Game.Settings.Server.NatPortMappingLifetime;
+			mapping = new Mapping(Protocol.Tcp, listen, external, lifetime, "OpenRA");
+			try
+			{
+				await natDevice.CreatePortMapAsync(mapping);
```

Comment:
Why not use the none async versions when we are using it with wait() right now?

Reply:
Actually, just because upstream demonstrated it that way https://github.com/alanmcgovern/Mono.Nat/blob/4c724791988afbfd1921a25896a2359e29532c9d/Mono.Nat.Console/Main.cs#L87